src/JustSaying/AwsTools/Publishing/MessagePublisherProvider.cs

```diff
@@ -0,0 +1,156 @@
+            _serializationRegister = serializationRegister;
+            _loggerFactory = loggerFactory;
+            _config = config;
+        }
+
+        ///<inheritdoc/>
+        public IMessagePublisher GetSnsPublisher(string topicName, bool throwOnPublishFailure, IDictionary<string, string> tags)
+        {
+            return GetTopicPublisher(topicName, throwOnPublishFailure, tags);
+        }
+
+        ///<inheritdoc/>
+        public IMessagePublisher GetSqsPublisher(string queueName, int retryCountBeforeSendingToErrorQueue)
+        {
+            return GetQueuePublisher(queueName, _config.Region, retryCountBeforeSendingToErrorQueue);
+        }
+
+        ///<inheritdoc/>
+        public ITopicCreator GetSnsCreator(string topicName, bool throwOnPublishFailure, IDictionary<string, string> tags)
+        {
+            return GetTopicPublisher(topicName, throwOnPublishFailure, tags);
+        }
+
+        ///<inheritdoc/>
+        public IQueueCreator GetSqsCreator(
+            string queueName,
+            string region,
+            int retryCountBeforeSendingToErrorQueue,
+            Dictionary<string, string> tags)
+        {
+            return GetQueuePublisher(queueName, region, retryCountBeforeSendingToErrorQueue);
+        }
+
+        private SqsPublisher GetQueuePublisher(string queueName, string region, int retryCountBeforeSendingToErrorQueue)
+        {
+            _writeLock.EnterUpgradeableReadLock();
+
+            try
+            {
+                if (_queuePublishers.ContainsKey(queueName))
+                {
+                    return _queuePublishers[queueName];
+                }
+                else
+                {
+                    try
+                    {
+                        _writeLock.EnterWriteLock();
+
+                        if (_queuePublishers.ContainsKey(queueName))
+                        {
+                            return _queuePublishers[queueName];
+                        }
+
+                        var regionEndpoint = RegionEndpoint.GetBySystemName(region);
+                        var client = _proxy.GetAwsClientFactory().GetSqsClient(regionEndpoint);
+
+                        _queuePublishers[queueName] = new SqsPublisher(regionEndpoint, queueName, client, retryCountBeforeSendingToErrorQueue, _serializationRegister, _loggerFactory)
+                        {
+                            MessageResponseLogger = _config.MessageResponseLogger
+                        };
+
+                        return _queuePublishers[queueName];
+                    }
+                    finally
+                    {
+                        _writeLock.ExitWriteLock();
+                    }
+                }
+            }
+            finally
+            {
+                _writeLock.ExitUpgradeableReadLock();
+            }
+        }
+
+        private SnsTopicByName GetTopicPublisher(string topicName, bool throwOnPublishFailure, IDictionary<string, string> tags)
```

Comment:
throwOnPublishFailure isn't used but is passed down from each method.
Was suppose to be used?

Reply:
Thanks! When I started this work I removed the `SnsWriteConfiguration.HandleException` which is a `Func<Exception, Message, bool>` on whether to throw an exception on publish, and replaced it with this parameter. I decided to not do that and put it in a separate PR, but this got left behind.

Removed now, thanks!